streaming/text/convert/laion/laion400m/convert_and_upload.py

```diff
@@ -0,0 +1,276 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Convert and upload LAION-400M parquet shards."""
+
+import json
+import os
+from argparse import ArgumentParser, Namespace
+from time import sleep, time
+from typing import Iterator, List, Optional, Union
+
+import numpy as np
+from pyarrow import parquet as pq
+
+from streaming import MDSWriter
+
+
+def parse_args() -> Namespace:
+    """Parse command-line arguments.
+
+    Returns:
+        Namespace: Command-line arguments.
+    """
+    args = ArgumentParser()
+    args.add_argument('--local',
+                      type=str,
+                      required=True,
+                      help='Local directory containing downloaded shards in parquet format.')
+    args.add_argument('--remote',
+                      type=str,
+                      required=True,
+                      help='Remote path to upload MDS-formatted shards to.')
+    args.add_argument('--keep_parquet',
+                      type=int,
+                      default=1,
+                      help='Whether to keep the parquet shards after conversion (about 10TB).')
+    args.add_argument('--keep_mds',
+                      type=int,
+                      default=1,
+                      help='Whether to keep the MDS shards after upload (about 10TB).')
+    args.add_argument('--hashes',
+                      type=str,
+                      default='sha1,xxh64',
+                      help='Hashes for validating shards, if any.')
+    args.add_argument('--poll_interval',
+                      type=float,
+                      default=30,
+                      help='Interval between polling for newly downloaded shards to process.')
+    return args.parse_args()
+
+
+def each_downloaded_shard(local: str) -> Iterator[int]:
+    """Iterate over each downloaded shard.
+
+    Args:
+        local (str): Local directory containing shards.
+
+    Returns:
+        Iterator[int]: Each downloaded shard ID.
+    """
+    basenames = set(os.listdir(local))
+    count = len(list(filter(lambda s: s.endswith('_stats.json'), basenames)))
+    for idx in range(count):
+        stats_filename = os.path.join(local, f'{idx:05}_stats.json')
+        assert os.path.exists(stats_filename)
```

Comment:
Please avoid `assert` statement in the src code. Better way to raise an Exception. Thanks!

Reply:
done